Add `data all` option to `push` to copy all annex'ed files

lschr opened the pull request:
This uses `git annex copy --all` to transfer all annexed data, not only files from current HEAD.

Useful for full data backup. 

- Fixes #7738.

lschr committed:
Add `data="all"` option to `push`

This uses `git annex copy --all` to transfer all annexed data, not only
files from current HEAD.

yarikoptic reviewed (commented):
some initial questions

yarikoptic commented:
@lschr I have for now converted back to draft. See comments above. Best is to rebase since CI testing was fixed up so should not have false positives for you any more.

lschr commented on datalad/core/distributed/push.py:
```diff
@@ -234,17 +237,43 @@ def __call__(
-            recursion_limit)
+            non_ds_paths = sorted(set(paths) - set(map(Path, subds)))
+            if non_ds_paths:
+                b = "\n".join(map(str, non_ds_paths))
```
It's been quite some time since I've worked on this, but I think `non_ds_paths` can have at most as many entries as the `path` parameter to this method. In this case, this can only be excessively long if the user passes a long list of paths. Is this something we have to worry about?

yarikoptic replied:
yes and the answer was yes ;)

this is an exception message. If it is too long, it could flood the screen/logs and hide away what matters -- actual message. Thus worth limiting to e.g. number of paths + sample path.

I think better to use regular string here for the format. Overall could be a message like what I propose for multiline diff (unverified)

lschr commented on datalad/core/distributed/tests/test_push.py:
```diff
@@ -259,6 +259,245 @@ def test_push(annex):
 
 
+@pytest.mark.parametrize("annex", [False, True])
+def test_push_all(annex, tmp_path):
```
To make sure that no regressions were introduced in non-annex mode.

Should I remove this?

lschr committed:
Push.__call__: Rename local variable for consistency

lschr commented:
Rebased on current `maint` and force-pushed.

lschr committed:
push command: Shorten error message if invalid paths are supplied

lschr committed:
push command: Use @with_tree in tests for `data="all"`

As @with_tree seems to be incompatible with pytest fixtures, don't run
test_push_all with annex=False.

lschr committed:
Better document the tests run in test_push_all_recursive

lschr committed:
test_push_all_recurvsive: Create only one subsubdataset per subdataset

Test execution time is reduced from 70 to 40 seconds on my laptop.

lschr commented:
Implemented suggestions and rebased on current `maint`.
